File: Data Structures & Algorithms/partition-equal-subset-sum/submission-4.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        S=sum(nums)
        n=len(nums)
        if S%2!=0:
            return False
        nums.sort()
        visited=set()
        def dfs(index,s):
            visited.add(index)
            if nums[index]+s == S//2:
                return S//2
            if nums[index]+s > S//2:
                visited.remove(index)
                return s
            cur=0
            i=index+1
            while i < n and cur!=S//2:
                cur=dfs(i,s+nums[index])
                i+=1
            if cur==S//2:
                return S//2
            else:
                visited.remove(index)
                return s
        if dfs(0,0)==S//2:
            indexes=[i for i in range(n) if i not in visited]
            if dfs(indexes[0],0)==S//2:
                return True
            else:
                return False
        return False

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-4.cpp
class Solution {
public:
    map<int,int> d ;
    void dfs(int index,vector<int>& nums){
        if(index <0) return ;
        int n= nums.size();
        int val=nums[index];
        if(d.find(val) != d.end()){
            int i = index+1;
            while(val != nums[i]){
                if(val < nums[i])
                   d[val] = max(d[val],d[nums[i]]+1);
                i++;
            }
        }
        else{
            int count =0;
            for(int i=index+1;i<n;i++){
                if(val<nums[i])
                    count = max(count,d[nums[i]]);
            }
            d[val]=count+1;
        }
        dfs(index-1,nums);
    }
    int lengthOfLIS(vector<int>& nums) {
        dfs(nums.size()-1,nums);
        int m=0;
        for (auto &[key, value] : d) {
            if(value > m)
               m=value ;
        }
        return m;
    }
};
